system-addon/content-src/lib/init-store.js

```diff
@@ -44,12 +44,25 @@ const messageMiddleware = store => next => action => {
 };
 
 const rehydrationMiddleware = store => next => action => {
+  if (!store._earlyMessageQueue) {
+    store._earlyMessageQueue = [];
+  }
+
   if (store._didRehydrate) {
     return next(action);
   }
 
   const isMergeStoreAction = action.type === MERGE_STORE_ACTION;
   const isRehydrationRequest = action.type === at.NEW_TAB_STATE_REQUEST;
+  const isSessionPerfAction = action.type === at.SAVE_SESSION_PERF_DATA;
+
+  // Treasure up the session perf actions to the early message queue if the hydration
```

Comment:
not sure what you mean by "Treasure up" ?

Reply:
Oops, my bad, what I was thinking here. Perhaps got influenced by this https://youtu.be/vm1GJMp0QN4?t=2681